sdks/python/apache_beam/transforms/core.py

```diff
@@ -1316,6 +1323,146 @@ def Map(fn, *args, **kwargs):  # pylint: disable=invalid-name
   return pardo
 
 
+def MapTuple(fn, *args, **kwargs):  # pylint: disable=invalid-name
+  """:func:`MapTuple` is like :func:`Map` but expects tuple inputs and
+  flattens them into multiple input arguments.
+
+      beam.MapTuple(lambda a, b, ...: ...)
+
+  is equivalent to Python 2
+
+      beam.Map(lambda (a, b, ...), ...: ...)
+
+  In other words
+
+      beam.MapTuple(fn)
+
+  is equivalent to
+
+      beam.Map(lambda element, ...: fn(*element, ...))
+
+  This can be useful when processing a PCollection of tuples
+  (e.g. key-value pairs).
+
+  Args:
+    fn (callable): a callable object.
+    *args: positional arguments passed to the transform callable.
+    **kwargs: keyword arguments passed to the transform callable.
+
+  Returns:
+    ~apache_beam.pvalue.PCollection:
+    A :class:`~apache_beam.pvalue.PCollection` containing the
+    :func:`MapTuple` outputs.
+
+  Raises:
+    ~exceptions.TypeError: If the **fn** passed as argument is not a callable.
+      Typical error is to pass a :class:`DoFn` instance which is supported only
+      for :class:`ParDo`.
+  """
+  if not callable(fn):
+    raise TypeError(
+        'MapTuple can be used only with callable objects. '
+        'Received %r instead.' % (fn))
+
+  if _fn_takes_side_inputs(fn):
+    wrapper = lambda x, *args, **kwargs: [fn(*(tuple(x) + args), **kwargs)]
+  else:
+    wrapper = lambda x: [fn(*x)]
+
+  label = 'MapTuple(%s)' % ptransform.label_from_callable(fn)
+
+  # Proxy the type-hint information from the original function to this new
+  # wrapped function.
+  get_type_hints(wrapper).input_types = get_type_hints(fn).input_types
+  output_hint = get_type_hints(fn).simple_output_type(label)
+  if output_hint:
+    get_type_hints(wrapper).set_output_types(typehints.Iterable[output_hint])
+
+  argspec = getfullargspec(fn)
+  num_defaults = len(argspec.defaults or ())
+  if num_defaults < len(args) + len(kwargs):
+    raise TypeError('Side inputs must have defaults for MapTuple.')
+  modified_argspec = inspect.ArgSpec(
```

Comment:
The code here is not Py3 compatible (see test failures). Could we reuse helpers in decorators.py or introduce a new helper there instead of calling inspect.* methods directly?
Also, we should keep in mind that we eventually want to support keyword-only arguments.

Reply:
I had played around with Python 3, but apparently that didn't make it into my final solution... Fixed.